File: go/pythagorean-triplet/pythagorean_triplet.go
// Package pythagorean is.
package pythagorean

import "math"

type Triplet [3]int

// Range returns a list of all Pythagorean triplets with sides in the
// range min to max inclusive.
func Range(min, max int) []Triplet {
	var ret []Triplet

	for a := min; a <= max; a++ {
		for b := a; b <= max; b++ {
			c := int(math.Sqrt(float64(a*a + b*b)))
			if c >= min && c <= max && isPythag(a, b, c) {
				ret = append(ret, Triplet{a, b, c})
			}
		}
	}

	return ret
}

func isPythag(a, b, c int) bool {
	return a*a + b*b == c*c
}

// Sum returns a list of all Pythagorean triplets where the sum a+b+c
// (the perimeter) is equal to p.
func Sum(p int) []Triplet {
	var ret []Triplet

	for b := p; b >= p / 3; b-- {
		for a := 1; a <= b; a++ {
			c := p - a - b
			if isPythag(a, b, c) {
				ret = append(ret, Triplet{a, b, c})
			}
		}
	}

	return ret
}

File: python/pythagorean-triplet/pythagorean_triplet.py
import math

def triplets_with_sum(number):
  triplets = {}
  for a in range(1, number):
    for b in range(a, number - a):
      c = number - a - b
      if is_triplet((a, b, c)):
        triplets.add((a, b, c))
  return triplets


def triplets_in_range(start, end):
  triplets = {}
  for a in range(start, end):
    for b in range(a, end - a):
      c = math.sqrt(a*a + b*b)
      if c != int(c):
        continue
      if start <= sum([a, b, c]) <= end:
        triplets.add((a, b, c))
  return triplets


def is_triplet(triplet):
  a, b, c = triplet
  return a*a + b*b == c*c


# vim:ts=2:sw=2:expandtab
